File: semgrep-container/test_vulnerabilities.js
// test_vulnerabilities.js

const fs = require('fs');
const child_process = require('child_process');
const crypto = require('crypto');
const express = require('express');
const bodyParser = require('body-parser');
const xml2js = require('xml2js');

const app = express();
app.use(bodyParser.json());

// 1. Eval-based RCE
function unsafeEval(userInput) {
  return eval(userInput);
}

// 2. Exec with unsanitized input
function listDir(dir) {
  child_process.exec(`ls ${dir}`, (err, stdout) => {
    console.log(stdout);
  });
}

// 3. Shell injection via execSync
function pingHost(host) {
  child_process.execSync(`ping -c 1 ${host}`, { stdio: 'inherit' });
}

// 4. Prototype pollution
function setProp(obj, key, value) {
  obj[key] = value;        // no check against "__proto__"
}

// 5. XSS: unescaped user input in HTML
app.get('/greet', (req, res) => {
  const name = req.query.name || 'guest';
  res.send(`<h1>Hello, ${name}!</h1>`);
});

// 6. Hardcoded credentials
const ADMIN_USER = 'admin';
const ADMIN_PASS = 'P@ssw0rd!';

// 7. Weak hashing (MD5)
function hashPassword(pw) {
  return crypto.createHash('md5').update(pw).digest('hex');
}

// 8. Insecure JWT secret
const jwtSecret = 'supersecretkey123';  // should come from env

// 9. XML External Entity (XXE)
app.post('/upload-xml', (req, res) => {
  const parser = new xml2js.Parser({                          // default allows XXE
    explicitArray: false
  });
  parser.parseString(req.body.xml, (err, result) => {
    res.json(result);
  });
});

// 10. Path traversal via fs.readFile
app.get('/read', (req, res) => {
  const file = req.query.file;  
  fs.readFile(`/var/data/${file}`, 'utf8', (err, data) => {
    res.send(err ? err.message : data);
  });
});

app.listen(3000, () => console.log('JS app listening on 3000'));

File: semgrep-container/test_vulnerabilities.py
# test_vulnerabilities.py

import os
import subprocess
import sqlite3
import pickle
import random
import hashlib
from flask import Flask, request

# 1. Eval-based RCE
def insecure_eval(user_input):
    # unsafely evaluating arbitrary Python code
    return eval(user_input)


# 2. Exec-based RCE
def insecure_exec(code):
    # executes arbitrary code strings
    exec(code)


# 3. Shell injection via os.system()
def list_files(path):
    # concatenation allows injection (e.g. path="; rm -rf /")
    os.system("ls " + path)


# 4. Subprocess with shell=True
def ping_host(host):
    # shell=True + string concat → command injection
    subprocess.call("ping -c 1 " + host, shell=True)


# 5. SQL Injection
def get_user(username):
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    # f-string without parameterization is injectable
    query = f"SELECT * FROM users WHERE name = '{username}'"
    cursor.execute(query)
    return cursor.fetchall()


# 6. Hardcoded credentials
ADMIN_USERNAME = "admin"
ADMIN_PASSWORD = "P@ssw0rd123"   # should never live in source


# 7. Weak hashing (MD5)
def hash_password(password):
    # MD5 is considered cryptographically broken
    return hashlib.md5(password.encode()).hexdigest()


# 8. Predictable “random” tokens
def generate_token():
    # random.randint is not secure for tokens
    return str(random.randint(1000, 9999))


# 9. Insecure deserialization
def load_session(data):
    # pickle.loads can execute arbitrary object constructors
    return pickle.loads(data)


# 10. Path traversal
def read_file(filename):
    # concatenation + no sanitization → attacker can use “../” to escape
    with open('/var/www/data/' + filename, 'r') as f:
        return f.read()


# --- minimal Flask app exposing one of the above flaws ---
app = Flask(__name__)
app.debug = True   # debug mode exposes internals

@app.route('/download')
def download():
    # uses read_file without validation → path traversal
    filename = request.args.get('file', '')
    return read_file(filename)

if __name__ == '__main__':
    # runs on all interfaces in debug mode
    app.run(host='0.0.0.0', port=5000)

